add tablayout for markdown file

parneet-guraya opened the pull request:
## Related Issues
App: #4025 

Library PR (if needed):

- [ ] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

## QA


https://github.com/owncloud/android/assets/111801812/85dea132-f6fd-41ed-bfc3-5c38eb0e2a49

parneet-guraya commented:
Hi, I have used tablayout now. I can also look into adding a viewpager if you suggest. Let me know what you think about it and I'm open to changes. :grinning:

JuancaG05 commented:
Hey @parneet-guraya! This looks really cool! It would be great if you add a ViewPager as you suggest so that we can have the gesture for switching between tabs. Once you add it I will take a look and do the code review! Thanks a lot 😀🥇

parneet-guraya commented:
> Hey @parneet-guraya! This looks really cool! It would be great if you add a ViewPager as you suggest so that we can have the gesture for switching between tabs. Once you add it I will take a look and do the code review! Thanks a lot grinning1st_place_medal

Hi, I'm looking into the viewpager functionality. These are the things I'm thinking to do for this. 

- Creating separate layout by extracting out `ScrollView (with direct child TextView)`.
- Then inflating the layout programatically if the file is non `.md` type.
- Otherwise programatically inflating `TabLayout` and `ViewPager2`. Then using `RecyclerView Adapter` inflate two layouts (one for markdown and another for ascii).
- Rest setting up the functionality together.

I just want the opinion on the layout and views inflating approach.
Thanks!

JuancaG05 commented:
Hi @parneet-guraya! It could work the way you suggest 😀. The only thing I can suggest there is not using a RecyclerView adapter, it's not necessary in this case, you can use a `FragmentStateAdapter` which is simpler. You can check more info and an example here: https://developer.android.com/guide/navigation/advanced/swipe-view-2

parneet-guraya commented:
> Hi @parneet-guraya! It could work the way you suggest 😀. The only thing I can suggest there is not using a RecyclerView adapter, it's not necessary in this case, you can use a `FragmentStateAdapter` which is simpler. You can check more info and an example here: https://developer.android.com/guide/navigation/advanced/swipe-view-2

Yes you're right `FragmentStateAdapter` will be better and I already looked into it and it was handling configuration changes as well!

The only thing I want to ask you is what fragment instance should I inflate in the adapter? 

Should I create separate fragment file. With the layout having `ScrollView (direct child TextView)`.

Then in the `PreviewTextFragment` file, If format is `.md` then setup `Viewpager2` with adapter(with two inflated fragment layouts) and attach with `TabLayout`.

Else create only single instance of the fragment and show it using fragment container view.

Also, for the `.md` case. I think adding `ViewPager2` and `TabLayout` in the layout is not a good idea( because they are gonna inflate everytime). We only want these when certain condition satisfy. Rather thinking to add programmatically when condition meets.

And for other files, maybe adding `FragmentContainer` (also programmatically) and then use it show the fragment.

Also we get the type of file in the onStart() callback, So this is where we will be checking (condition) and inflating respective views. But, as user will go to home screen and comeback. would it be bad performance wise? As views will be reinflating everytime user go back and forth.

Let me know about your opinion and then I will start working on that 🚀
Thanks!!

JuancaG05 commented:
Since we are using the same fragment, in this case `PreviewTextFragment`, to show both normal text and Markdown files, I guess the best option is inflating views programmatically as you suggest, yeah. Maybe we can analyse this in a future and separate it in two different fragments if we see it's simpler and more effective, but for the moment, let's keep that approach.

About the `onStart` callback you comment, you can try moving it to `onCreate`, but I think this is currently in `onStart` because to execute the async task of loading the text we need the fragment to be already created and the context in a stable state. Anyways, don't worry too much about it, although it inflates every time, it's not such a complex layout and I don't think it's a common case that the user switches between home screen and the app repeatedly 😀

parneet-guraya commented:
> Since we are using the same fragment, in this case `PreviewTextFragment`, to show both normal text and Markdown files, I guess the best option is inflating views programmatically as you suggest, yeah. Maybe we can analyse this in a future and separate it in two different fragments if we see it's simpler and more effective, but for the moment, let's keep that approach.
> 
> About the `onStart` callback you comment, you can try moving it to `onCreate`, but I think this is currently in `onStart` because to execute the async task of loading the text we need the fragment to be already created and the context in a stable state. Anyways, don't worry too much about it, although it inflates every time, it's not such a complex layout and I don't think it's a common case that the user switches between home screen and the app repeatedly grinning

Okay, got it. Making sure we're on the same page. So, I will be creating a different fragment with only having `ScrollView with TextView` and use that  fragment for either of the cases `(.md/other types)` to show text inside `PreviewTextFragment` Fragment

JuancaG05 commented:
> Okay, got it. Making sure we're on the same page. So, I will be creating a different fragment with only having `ScrollView with TextView` and use that fragment for either of the cases `(.md/other types)` to show text inside `PreviewTextFragment` Fragment

Yes, but maybe not even necessary another fragment but a XML layout that you can inflate wherever 😄 (in this case `PreviewTextFragment`). You can even keep the current one (`preview_text_fragment.xml`) to show normal texts and create a new one that already contains the tabs and the ViewPager, so that you can inflate the whole layout programmatically instead of the each UI element separately 👍

parneet-guraya commented:
Yes, I can create a separate xml layout with having viewpager and TabLayout. But, about the inflating part `FragmentStateAdapter` needs to create a new Fragment instance if we're using it. So, we cannot inflate xml layout. What do you think about it?

JuancaG05 commented:
Check: https://github.com/android/views-widgets-samples/tree/master/ViewPager2. There you can see different examples of tabs inflating whole fragments or just some views. These examples use RecyclerView adapter though 😞, but just to see if it brings some ideas.

If you finally need to create a new Fragment class so that you can return it from the adapter, I suggest to create a private class in the same class of the adapter, since this is not going to be used from anywhere else at the moment.

parneet-guraya commented:
![Screenshot from 2023-06-27 17-18-26](https://github.com/owncloud/android/assets/111801812/083f68a9-b08d-476f-be73-6a0a20a1c7d9)

Internally this is what `FragmentStateAdapter` is doing it has a RecyclerView Adapter with implementing StatefulAdapter which facilitates saving state. So, It is said in the github page if we're using simple views with recyclerview adapter we can implement `StatefulAdapter` interface to handle state of the view.

So, if we are inflating a layout, along with recyclerview adapter we also need to implement the interface and maybe replicate the behaviour `FragmentStateAdapter` is doing which is described in the picture I attached. Now knowing this which approach should be the best bet?

JuancaG05 commented:
I would go anyways with `FragmentStateAdapter`, who knows if it will grow in a future 😀. Don't be afraid of making mistakes coding here, just go for it and if you need to change something you can do it, that's also a great way to learn 😁

parneet-guraya commented:
Added new commit, What does it do? ->

- [x] Show TabLayout with ViewPager for `.md` file type (Tabs -> Markdown & Text). Also handles configuration changes well.
- [x] Showing normal TextView for non-`.md` files. Also handles configuration changes.

- I was having trouble with programatically created views (as they were not handling configuration/state changes). So, all the views are inflated and later removed when not needed. Although, I know it's not the best approach to do, let me know any workround if this approach is feasible.

- Also I moved `loadAndShowTextPreview()` from `onStart()` to `onViewCreated()` which saves the unnecessary inflation and also saves us from handling state saving/restoration.

You can review this and let me know if changes are needed in this approach.
Thanks!


https://github.com/owncloud/android/assets/111801812/f32a2f52-d1ed-4b9c-8e27-6de654d3f2d4

JuancaG05 commented on owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewTextFragment.java:
```diff
@@ -173,6 +181,12 @@ public void onActivityCreated(Bundle savedState) {
         mProgressController.setProgressBar(mProgressBar);
     }
 
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        loadAndShowTextPreview();
```
🥇

JuancaG05 commented on owncloudApp/src/main/res/values/strings.xml:
```diff
@@ -758,5 +758,7 @@
     <string name="ocis_accounts_warning_message">Please, remove the account and login again to get the spaces feature</string>
     <string name="ocis_accounts_warning_checkbox_message">Don\'t show again</string>
     <string name="ocis_accounts_warning_button">Understood</string>
+    <string name="tab_label_markdown">Markdown</string>
+    <string name="tab_label_ascii">Text</string>
```
Ah! Indeed you added them here! It's just that you didn't use them 😄. Use them in the code and move the "Markdown" string to the non-translatable ones 😁

JuancaG05 reviewed (changes requested):
Hey @parneet-guraya, veeeery cool job! 💯 I just added some comments to improve a bit your work but, overall it's very good! Congrats! 🎖️

parneet-guraya commented on owncloudApp/src/main/java/com/owncloud/android/ui/preview/PreviewFormatTextFragmentStateAdapter.kt:
```diff
@@ -0,0 +1,103 @@
+import io.noties.markwon.html.HtmlPlugin
+
+class PreviewFormatTextFragmentStateAdapter(
+    fragment: Fragment, private val text: String, private val mimeType: String
```
Hi, yes it's not being used anywhere outside. But it isn't being exposed either, because it's not a property and we can access value parameters (as [kotlin docs](https://kotlinlang.org/docs/classes.html#constructors:~:text=Primary%20constructor%20parameters%20can%20be%20used%20in%20the%20initializer%20blocks.%20They%20can%20also%20be%20used%20in%20property%20initializers%20declared%20in%20the%20class%20body%3A) says) of a primary constructor inside init block as well as class level property initializer (in this case it would be:
```
val formatTypes =
        mapOf(TYPE_MARKDOWN to fragment.getString(R.string.tab_label_markdown), TYPE_PLAIN to fragment.getString(R.string.tab_label_ascii))
```
So, It's just like a local parameter of constructor (like in java). Hence I cannot mark it as private and if I want to it needs to be a property which in fact it's not being used as.
- If I set it private only, IDE (error) -> `Constructor parameter is never used as a property` 
- and if i set it `private val ..`, IDE (warning with `val` grayed out) -> Constructor parameter is never used as a property

JuancaG05 replied:
Then we'll follow Kotlin rules! 😀

parneet-guraya commented:
Hi, I've added the changes you requested. Also, as you can see in the Adapter file, I've defined it in a way that if in future we want to support some other format also in a same manner (with two tabs) we can do it as we're passing mimeType in the adapter and we can handle it in the fragment it is creating and show accordingly. What do you think about this approach that i've used (and not going with tightly coupled approach that will only work for Markdown)?

Also I've some questions about the `PreviewTextFragment` file ->
- As we're using ViewModels inside this, but we're not using `ViewModelProvider` instead we're using `get()`. According to current approach the ViewModel instance would not be cached and will not be passed when it is reinitialized (let's say due to configuration change) ,with the same instance(viewmodel) instead it would create a new ViewModel instance everytime (because we're not using ViewModelProvider since using it will automatically passes the cached instance). Is there any particular reason for that?
- Also what about converting it to kotlin and maybe using Coroutines (though i'm very beginner to this but it is the recommended and new approach) instead of AsyncTask?

Thanks!

JuancaG05 commented:
Hi @parneet-guraya! About the uncoupled adapter, that's perfect! Pretty nice solution, you are a pro! 💯 

About the ViewModel, don't worry about that because this is a temporary solution until we move this class to Kotlin and we can inject the ViewModel properly, as we do in other classes. We had to add the ViewModel some weeks ago and the quicker approach was this, taking into account that we have in our roadmap the kotlinization of these classes 😉

And yes, your second question connects with my previous answer, we have to move the whole class to Kotlin and AsyncTask will be replaced by coroutines as you suggest. But this is out of the scope of this issue, we'll work on it in a future 😀

JuancaG05 reviewed (approved):
Very nice job @parneet-guraya! Thanks a lot for this valuable contribution! We hope to see you again over here 😸🥇

jesmrec commented:
After the code review is approved, now the final QA phase starts to validate the functional feature. I have to thank your engagement in the code review @parneet-guraya, that was awesome.

Feature works very fine and smooth, i detected just one detail, let me know if it is something feasible or fixable:

Content displayed is pretty tight with the tabs, maybe a short space here will make it even better visible:

![Screenshot 2023-07-10 at 15 07 27](https://github.com/owncloud/android/assets/14894746/c610f83b-d460-4e38-ace8-a9201febc56b)

Would a fix affect to all the .txt files? 

Let me know your opinion about!

parneet-guraya commented:
Hi, @jesmrec  As I have inflated the same layout for simple `.txt` and `.md` (in adapter). Maybe I can add top margin to the Scroll View 🤔? What's your opinion about it?

parneet-guraya commented:
## Before
| Markdown | Text  |
|-------------|-------------|
| ![Before_markdown](https://github.com/owncloud/android/assets/111801812/ec0a4cc0-5247-4e3e-b0a8-8cc894f5cec9) | ![Before_text](https://github.com/owncloud/android/assets/111801812/7566ac37-dd90-4067-b9c4-4b6d1d0391ec) |

## After
>Would a fix affect to all the .txt files?

Yes, As I am using same layout for showing `.txt` and `.md` both. But, I think `.txt` file also needed some space and are looking fine to me.

I have added `padding` to `TextView` `Top` and `Bottom` with size `standard_half_margin (8dp)`. Bottom space was looking tight so added that as well.

Let me know you opinion about the change :smile:.
| Markdown | Text  |
|-------------|-------------|
| ![After_markdown](https://github.com/owncloud/android/assets/111801812/166378ca-f66a-4af1-ba7e-8c015573a53a) | ![After_text](https://github.com/owncloud/android/assets/111801812/736d52fe-c5ae-49d8-9d1d-71b5de9d3804) |

jesmrec commented:
yeah! that's good so the padding makes the text to be placed better. That works for me.

From my side this is approved and ready to be merged.

Thanks a lot for the contribution @parneet-guraya ! great job!
